trunk/source/Library-UltimateCore/src/de/uni_freiburg/informatik/ultimate/core/lib/models/annotation/CheckNegativeMessageProvider.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2023 Manuel Bentele
+ *
+ * This file is part of the ULTIMATE Core.
+ *
+ * The ULTIMATE Core is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ULTIMATE Core is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ULTIMATE Core. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Additional permission under GNU GPL version 3 section 7:
+ * If you modify the ULTIMATE Core, or any covered work, by linking
+ * or combining it with Eclipse RCP (or a modified version of Eclipse RCP),
+ * containing parts covered by the terms of the Eclipse Public License, the
+ * licensors of the ULTIMATE Core grant you additional permission
+ * to convey the resulting work.
+ */
+
+package de.uni_freiburg.informatik.ultimate.core.lib.models.annotation;
+
+import de.uni_freiburg.informatik.ultimate.core.lib.models.annotation.Check.Spec;
+
+/**
+ * Negative default message provider for {@link Check} specifications that are checked.
+ * 
+ * @author Manuel Bentele
+ */
+public class CheckNegativeMessageProvider extends CheckMessageProvider {
+
+	@Override
+	public String getDefaultMessage(final Spec spec) {
+		switch (spec) {
+		case ARRAY_INDEX:
```

Comment:
I like your refactoring there, this moves these huge switch-cases out of check. However, one issue is that, if you want to add a new Spec, you need to change three files (`Check`, `CheckPositiveMessageProvider`, `CheckNegativeMessageProvider`) and these switch-cases are still there (just somewhere else). So since you started this, is there a nicer solution? Maybe it is better if you can just add two methods to `Spec` for positive/negative message (along with a constructor)? This eliminates these switch-cases and you only need one class for `CheckMessageProvider`.

Reply:
You're right. If a new `Spec` should be added, three files have to be edited. Your proposed solution would be one possibility to get rid of the switch-case statements.